crd-generator/api/src/main/java/io/fabric8/crd/generator/v1/decorator/EnsureSingleStorageVersionDecorator.java

```diff
@@ -19,46 +19,39 @@
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionSpecFluent;
 import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionVersion;
-import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionVersionBuilder;
 
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Predicate;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class EnsureSingleStorageVersionDecorator
     extends CustomResourceDefinitionDecorator<CustomResourceDefinitionSpecFluent<?>> {
 
-  private final AtomicReference<String> storageVersion = new AtomicReference<>();
-
   public EnsureSingleStorageVersionDecorator(String name) {
     super(name);
   }
 
   @Override
   public void andThenVisit(CustomResourceDefinitionSpecFluent<?> spec, ObjectMeta resourceMeta) {
-    Predicate<CustomResourceDefinitionVersionBuilder> hasStorageVersion = v -> v.getStorage() != null && v.getStorage();
+    List<CustomResourceDefinitionVersion> versions = spec.buildVersions();
 
-    if (spec.hasVersions() && !spec.hasMatchingVersion(hasStorageVersion)) {
-      spec.editFirstVersion().withStorage(true).endVersion();
-    }
+    List<String> storageVersions = versions.stream()
+        .filter(v -> Optional.ofNullable(v.getStorage()).orElse(true))
```

Comment:
For endusers, it's already documented in the annotation itself:
https://github.com/fabric8io/kubernetes-client/blob/main/kubernetes-model-generator/kubernetes-model-common/src/main/java/io/fabric8/kubernetes/model/annotation/Version.java#L46

```java
/**
   * Whether this version corresponds to the persisted version for the associated CRD. Note that only one version can set
   * {@code storage} to {@code true} for a given CRD.
   *
   * @return {@code true} if this version corresponds to the persisted version for the associated CRD, {@code false} otherwise
   */
  boolean storage() default true; // <-- here
```


For (internal) CRDGenerator developers:
If a property like `storage` is equal to the default value, it is usually set to null in the internal representation.
This allows to omit the property during the schema generation, so that only those properties end up in the schema for which the value is not equal to the schema default.

We could start with a `DEVELOPMENT.md` file in the CRDGenerator package to document such internal behaviours. (There are a few more things which should be documented e.g. the sorting algorithm for the decorators...)

Reply:
> For endusers, it's already documented in the annotation itself: https://github.com/fabric8io/kubernetes-client/blob/main/kubernetes-model-generator/kubernetes-model-common/src/main/java/io/fabric8/kubernetes/model/annotation/Version.java#L46
> 
> ```java
> /**
>    * Whether this version corresponds to the persisted version for the associated CRD. Note that only one version can set
>    * {@code storage} to {@code true} for a given CRD.
>    *
>    * @return {@code true} if this version corresponds to the persisted version for the associated CRD, {@code false} otherwise
>    */
>   boolean storage() default true; // <-- here
> ```

Right, I missed that. This might be good enough.